altimeter_ms5607_spi/altimeter_ms5607_spi.cc

```diff
@@ -0,0 +1,142 @@
+    : _hspi(hspi), _csPort(csPort), _csPin(csPin), _misoPort(misoPort), _misoPin(misoPin), _seaLevelPressure(seaLevelPressure) {}
+
+bool AltimeterMs5607Spi::Reset() {
+    bool opStatus = true;
+
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_RESET);
+    // reset, opcode 0x1E
+    if (HAL_SPI_Transmit(_hspi, (uint8_t *)(const uint8_t[]){0x1E}, 1, SERIAL_TIMEOUT) != HAL_OK) opStatus = false;
+    // poll for reset complete (approx 2.8ms)
+    GPIO_PinState readStatus = GPIO_PIN_RESET;
+    while (readStatus == GPIO_PIN_RESET) {
+        readStatus = HAL_GPIO_ReadPin(_misoPort, _misoPin);
+    }
+    HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+    return opStatus;
+}
+
+AltimeterMs5607Spi::State AltimeterMs5607Spi::Init() {
+    _state = IDLE;
+
+    // read factory coefficients
+    for (int i = 0; i < 8; i++) {
+        uint8_t buffer[2];
+        HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_RESET);
+        // PROM read, opcode 0b1010xxx0
+        uint8_t command[1] = {(uint8_t)(0b10100000 | (i << 1))};
+        if (HAL_SPI_Transmit(_hspi, command, 1, SERIAL_TIMEOUT) != HAL_OK) _state = ERROR;
+        if (HAL_SPI_Receive(_hspi, buffer, 2, SERIAL_TIMEOUT) != HAL_OK) _state = ERROR;
+        _coefficients[i] = ((uint16_t)buffer[0] << 8) | (uint16_t)buffer[1];
+        HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+    }
+
+    // CRC algorithm copied from TE application note AN520
+    uint16_t crc = 0xF & _coefficients[7];
+    uint16_t crcActual = 0;
+    _coefficients[7] = 0xFF00 & _coefficients[7];
+    for (int i = 0; i < 16; i++) {
+        crcActual ^= i % 2 == 0 ? _coefficients[i >> 1] >> 8 : _coefficients[i >> 1] & 0x00FF;
+        for (int j = 0; j < 8; j++) {
+            crcActual = crcActual & 0x8000 ? (crcActual << 1) ^ 0x3000 : (crcActual << 1);
+        }
+    }
+    crcActual = crcActual >> 12;
+    if (crc != crcActual) _state = ERROR;
+
+    return _state;
+}
+
+AltimeterMs5607Spi::State AltimeterMs5607Spi::Convert(Rate rate) {
+    switch (_state) {
+        case IDLE: {
+            _state = POLL_D1;
+
+            HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_RESET);
+            // convert D1 pressure, the value of the rate enum is the opcode of the corresponding conversion
+            uint8_t command = (uint8_t)rate;
+            if (HAL_SPI_Transmit(_hspi, &command, 1, SERIAL_TIMEOUT) != HAL_OK) _state = ERROR;
+            break;
+        }
+
+        case POLL_D1: {
+            if (HAL_GPIO_ReadPin(_misoPort, _misoPin) == GPIO_PIN_RESET) break;
+            HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);
+
+            _state = POLL_D2;
+
+            uint8_t buffer[3];
+            HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_RESET);
+            // read data, opcode 0x00
+            if (HAL_SPI_Transmit(_hspi, (uint8_t *)(const uint8_t[]){0x00}, 1, SERIAL_TIMEOUT) != HAL_OK) _state = ERROR;
```

Comment:
if this command transmit fails, do we still want to attempt the receive on the next line?

Reply:
Ideally no, but I need the following line `HAL_GPIO_WritePin(_csPort, _csPin, GPIO_PIN_SET);` to properly terminate the communication. the chip gives undefined behavior otherwise